.github/workflows/ci.yml

```diff
@@ -8,9 +8,236 @@ on:
+  BENCH: 'yes'
   SECP256K1_BENCH_ITERS: 2
+  CTIMETESTS: 'yes'
+  # Compile and run the examples.
+  EXAMPLES: 'yes'
 
 jobs:
+  docker_wine_cache:
+    name: 'Populate Docker cache'
+    runs-on: ubuntu-latest
+    steps:
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+
+      - name: Build container
+        uses: docker/build-push-action@v4
+        with:
+          file: ./ci/linux-debian-wine.Dockerfile
+          tags: ci-image
+          push: false
+          load: false
+          cache-from: type=gha
+          cache-to: type=gha,mode=min
+
+  mingw_debian:
+    name: ${{ matrix.configuration.job_name }}
+    runs-on: ubuntu-latest
+    needs: docker_wine_cache
+
+    env:
+      WRAPPER_CMD: 'wine'
+      WITH_VALGRIND: 'no'
+      ECDH: 'yes'
+      RECOVERY: 'yes'
+      SCHNORRSIG: 'yes'
+      CTIMETESTS: 'no'
+
+    strategy:
+      fail-fast: false
+      matrix:
+        configuration:
+          - job_name: 'x86_64 (mingw32-w64): Windows (Debian stable, Wine)'
+            env_var: 'HOST=x86_64-w64-mingw32'
+          - job_name: 'i686 (mingw32-w64): Windows (Debian stable, Wine)'
+            env_var: 'HOST=i686-w64-mingw32'
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+
+      - name: Build container
+        uses: docker/build-push-action@v4
+        with:
+          context: .
+          file: ./ci/linux-debian-wine.Dockerfile
+          tags: ci-image
+          push: false
+          load: true
+          cache-from: type=gha
+
+      - name: Run test script
+        uses: addnab/docker-run-action@v3
+        with:
+          image: ci-image
+          options: -v ${{ github.workspace }}:${{ github.workspace }}
+          run: |
+            export \
+              HOST="${{ env.HOST }}" \
+              WRAPPER_CMD="${{ env.WRAPPER_CMD }}" \
+              WERROR_CFLAGS="${{ env.WERROR_CFLAGS }}" \
+              MAKEFLAGS="${{ env.MAKEFLAGS }}" \
+              BUILD="${{ env.BUILD }}" \
+              ECMULTWINDOW="${{ env.ECMULTWINDOW }}" \
+              ECMULTGENPRECISION="${{ env.ECMULTGENPRECISION }}" \
+              ASM="${{ env.ASM }}" \
+              WIDEMUL="${{ env.WIDEMUL }}" \
+              WITH_VALGRIND="${{ env.WITH_VALGRIND }}" \
+              EXTRAFLAGS="${{ env.EXTRAFLAGS }}" \
+              EXPERIMENTAL="${{ env.EXPERIMENTAL }}" \
+              ECDH="${{ env.ECDH }}" \
+              RECOVERY="${{ env.RECOVERY }}" \
+              SCHNORRSIG="${{ env.SCHNORRSIG }}" \
+              ELLSWIFT="${{ env.ELLSWIFT }}" \
+              SECP256K1_TEST_ITERS="${{ env.SECP256K1_TEST_ITERS }}" \
+              BENCH="${{ env.BENCH }}" \
+              SECP256K1_BENCH_ITERS="${{ env.SECP256K1_BENCH_ITERS }}" \
+              CTIMETESTS="${{ env.CTIMETESTS }}" \
+              EXAMPLES="${{ env.EXAMPLES }}" \
+              ${{ matrix.configuration.env_var }}
```

Comment:
I suspect this is necessary because `addnab/docker-run-action@v3`. I wonder if there's a better way or a different action we could use, or if we could just run docker manually.

But okay, yes, I am starting to see that this approach of using the GHA cache also has its disadvantages.

Reply:
> run docker manually

We could perhaps do something like this (assuming we add a unique prefix our env variables) https://stackoverflow.com/a/61955515

Reply:
[Reworked](https://github.com/bitcoin-core/secp256k1/pull/1398#issuecomment-1674068461).